pulsar-broker/src/main/java/org/apache/pulsar/broker/stats/prometheus/TopicStats.java

```diff
@@ -382,77 +389,102 @@ static void metricType(SimpleTextOutputStream stream, String name) {
-                .write("\"} ");
-        stream.write(value);
-        appendEndings(stream);
+    private static void metric(Map<String, List<String>> metrics, String cluster, String namespace, String topic,
+                               String subscription, String consumerName, long consumerId, String name, double value,
+                               boolean splitTopicAndPartitionIndexLabel) {
+        writeMetric(metrics, name, stream -> {
+            writeRequiredLabels(stream, cluster, namespace, topic, name, splitTopicAndPartitionIndexLabel);
+            stream.write("\",subscription=\"").write(subscription)
+                    .write("\",consumer_name=\"").write(consumerName).write("\",consumer_id=\"")
+                    .write(consumerId).write("\"} ")
+                    .write(value);
+            writeEndings(stream);
+        });
     }
 
-    private static void metric(SimpleTextOutputStream stream, String cluster, String namespace, String topic,
-            String subscription, String consumerName, long consumerId, String name, double value,
-            boolean splitTopicAndPartitionIndexLabel) {
-        metricType(stream, name);
-        appendRequiredLabels(stream, cluster, namespace, topic, name, splitTopicAndPartitionIndexLabel)
-                .write("\",subscription=\"").write(subscription)
-                .write("\",consumer_name=\"").write(consumerName).write("\",consumer_id=\"")
-                .write(consumerId).write("\"} ");
-        stream.write(value);
-        appendEndings(stream);
+    private static void metricWithRemoteCluster(Map<String, List<String>> metrics, String cluster, String namespace,
+                                                String topic, String name, String remoteCluster, double value,
+                                                boolean splitTopicAndPartitionIndexLabel) {
+        writeMetric(metrics, name, stream -> {
+            writeRequiredLabels(stream, cluster, namespace, topic, name, splitTopicAndPartitionIndexLabel);
+            stream.write("\",remote_cluster=\"").write(remoteCluster).write("\"} ").write(value);
+            writeEndings(stream);
+        });
     }
 
-    private static void metricWithRemoteCluster(SimpleTextOutputStream stream, String cluster, String namespace,
-            String topic, String name, String remoteCluster, double value, boolean splitTopicAndPartitionIndexLabel) {
-        metricType(stream, name);
-        appendRequiredLabels(stream, cluster, namespace, topic, name, splitTopicAndPartitionIndexLabel)
-                .write("\",remote_cluster=\"").write(remoteCluster).write("\"} ");
-        stream.write(value);
-        appendEndings(stream);
+    private static void writeMetric(Map<String, List<String>> metrics, String name,
+                                    Consumer<SimpleTextOutputStream> metricWriter) {
+        ByteBuf buf = ByteBufAllocator.DEFAULT.heapBuffer();
```

Comment:
it looks like we are doing more memory allocations (even if you call 'buf.release()') 
is it possible to move back to creating the SimpleTextOutputStream only once ?

Reply:
agree - have updated to use prometheus' `Collector.MetricFamilySamples` with only one `SimpleTextOutputStream` used when writing